zebrad/src/commands/start/sync.rs

```diff
@@ -0,0 +1,204 @@
+use color_eyre::Report;
+use eyre::eyre;
+use futures::stream::{FuturesUnordered, StreamExt};
+use std::{collections::HashSet, iter};
+use tower::{Service, ServiceExt};
+use tracing_futures::Instrument;
+use zebra_chain::{block::BlockHeaderHash, types::BlockHeight};
+
+pub struct Syncer<ZN, ZS>
+where
+    ZN: Service<zebra_network::Request>,
+{
+    pub peer_set: ZN,
+    pub state: ZS,
+    pub tip_requests: FuturesUnordered<ZN::Future>,
+    pub block_requests: FuturesUnordered<ZN::Future>,
+    pub block_locator: Vec<BlockHeaderHash>,
+    pub downloading: HashSet<BlockHeaderHash>,
+    pub downloaded: HashSet<BlockHeaderHash>,
+    pub fanout: NumReq,
+}
+
+impl<ZN, ZS> Syncer<ZN, ZS>
+where
+    ZN: Service<zebra_network::Request, Response = zebra_network::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    ZN::Future: Send,
+    ZS: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
+        + Send
+        + Clone
+        + 'static,
+    ZS::Future: Send,
+{
+    pub async fn run(&mut self) -> Result<(), Report> {
+        loop {
+            if self.tip_requests.is_empty() {
+                info!("populating prospective tips list");
+                self.obtain_tips(vec![super::GENESIS]).await?;
+            }
+
+            info!("extending prospective tips");
+            self.extend_tips().await?;
+            self.process_blocks().await?;
+        }
+    }
+    /// Given a block_locator list fan out request for subsequent hashes to
+    /// multiple peers
+    pub async fn obtain_tips(&mut self, block_locator: Vec<BlockHeaderHash>) -> Result<(), Report> {
+        for _ in 0..self.fanout {
+            let req = self.peer_set.ready_and().await.map_err(|e| eyre!(e))?.call(
+                zebra_network::Request::FindBlocks {
+                    known_blocks: block_locator.clone(),
+                    stop: None,
+                },
+            );
+            self.tip_requests.push(req);
```

Comment:
I don't think it works to store the tip futures in a `FuturesUnordered` for later processing like this, because we need to keep track of the different tips we're extending and combine the responses together to figure out what to do next.  We could wait on all of the fanout requests in parallel using a `FuturesUnordered` inside this function, but steps 2-5 of the `ObtainTips` description require that we have all of the responses in order to process and combine them.

Reply:
More precisely, if some peers are unreliable, we can stop after we have some of the responses, and we've given up on the other responses.